### PR TITLE
Produce WarningInstrumentingWithUprobesEvents

### PR DESCRIPTION
**Pull request by dpallotti**
This finally makes `LinuxTracing` produce `WarningInstrumentingWithUprobesEvent`
if necessary.
A helper function is added to `LinuxTracingUtils.h`/`.cpp` to list the functions
that uprobes cannot instrument, and for each to give an explanation that will be
displayed in the client. The return type is inspired by
`orbit_user_space_instrumentation::InstrumentationManager::InstrumentationResult::function_ids_to_error_messages`.
From the `std::map`, a `WarningInstrumentingWithUprobesEvent` is produced by
`orbit_linux_tracing::TracerImpl::Startup`.

Screenshot: http://screen/6WS59EVsNGBjYgB

Bug: http://b/232072696

Test:
- Unit tests.
- Try to instrument some functions with uprobes from the "unaligned"
  `triangle.exe` and verify the message in the "Capture Log".
- Instrument some functions with uprobes from Trata and verify that no warning
  is given.

### CODE REVIEW

File: src/LinuxTracing/LinuxTracingUtils.cpp [hunk 1]
```diff
@@ -213,4 +222,121 @@ bool SetMaxOpenFilesSoftLimit(uint64_t soft_limit) {
   return true;
 }
 
+// Check for the existence of a file mapping containing the absolute address of the function. We
```
Comment (florian-kuebler) [inline src/LinuxTracing/LinuxTracingUtils.cpp:225]: So we would only report functions that we can not instrument at all. Wouldn't it be nicer to report the once where we fail for some mappings? Otherwise, I might miss calls that I would expect.

Comment (dpallotti) [reply src/LinuxTracing/LinuxTracingUtils.cpp:225]: Hmm I didn't think about it (having a module mapped multiple times is a case I still don't fully comprehend). So you suggest to change this function to return true only if a file mapping containing the absolute address is found for all the `ModuleInfo`s?

Comment (florian-kuebler) [reply src/LinuxTracing/LinuxTracingUtils.cpp:225]: I don't think it would be restricted to only one file mapping.

I would change it to only return true for a given function, if all modules that contain this function are mapped as a file mapping.

Comment (dpallotti) [reply src/LinuxTracing/LinuxTracingUtils.cpp:225]: > I don't think it would be restricted to only one file mapping.

I didn't get this part, because from the next one I feel like we are saying the same thing. Could you clarify?

File: src/LinuxTracing/LinuxTracingUtils.cpp [hunk 2]
```diff
@@ -213,4 +222,121 @@ bool SetMaxOpenFilesSoftLimit(uint64_t soft_limit) {
...
+// have to consider the possibility that the module be mapped multiple times, and hence that the
+// function have multiple absolute addresses.
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply src/LinuxTracing/LinuxTracingUtils.cpp]: Hehe, it was subjunctive, on purpose because it follows "the possibility that". I guess nobody uses it anymore... I switched to adding "may" to avoid confusion.

File: src/LinuxTracing/LinuxTracingUtils.cpp [hunk 3]
```diff
@@ -213,4 +222,121 @@ bool SetMaxOpenFilesSoftLimit(uint64_t soft_limit) {
...
+//
+// Note: A more naive solution would be to look for a map containing the file offset for the
+// function, hence not involving absolute addresses and modules at all. For misaligned PEs, this can
+// cause false negatives, because a function can be mapped twice, in a file mapping and in an
+// anonymous (executable) mapping, but with the actual absolute address of the function being in the
+// anonymous (executable) mapping.
+//
+// Example: Consider a PE with one section, the .text section, at offset in the file 0x400 and
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply src/LinuxTracing/LinuxTracingUtils.cpp:235]: > Would it be sufficient to disregard the non executable segments to remove the ambiguity described below?

Conceptually, unfortunately no. In the example I show the file mapping corresponding to the headers, followed by the file mapping corresponding to the section containing the function. However, there could be a case where the PE has two consecutive executable sections, and the first one happens to be aligned "correctly" (such that a file mapping is possible) by chance: we could have the same problem, with the first map also containing part of the second section, and for a function in the second section the problem would be the same, but this time the first map is also executable.

In other words: there could be a pair of mappings like in the example below, but with the file mapping also corresponding to an executable section and hence executable.

I hope this makes sense.

[the remark this answers is not in the document]
Comment (dpallotti) [reply src/LinuxTracing/LinuxTracingUtils.cpp:235]: > Also it might be helpful to sort out the non-executable segements for another reason: The code below looks as if it would not complain if it finds a function only in a non-executable segment belonging to a file mapping.

This is true and I would say this is intended here. Basically here we are considering valid cases that are "interesting" in the context of uprobes: module not yet loaded at all (not supported by Orbit, yet?, but supported by uprobes), and part of the module loaded with an anonymous map (but uprobes needs a file mapping).

In this context, a function not in an executable section is not a valid case and is out of scope: it means that something is wrong at a different level: either we made some mistake somewhere in computing where the function would be in memory, or the module is not fully mapped and the function will not get any hit from any kind of dynamic instrumentation simply because it won't be executed. If we really want to check for this case, we need to do it somewhere else, in a different context.

I hope this also makes sense.

File: src/LinuxTracing/LinuxTracingUtils.cpp [hunk 4]
```diff
@@ -213,4 +222,121 @@ bool SetMaxOpenFilesSoftLimit(uint64_t soft_limit) {
...
+  for (const ModuleInfo& module : file_path_modules) {
+    ORBIT_CHECK(module.file_path() == function.file_path());
+    const uint64_t function_absolute_address =
+        orbit_module_utils::SymbolVirtualAddressToAbsoluteAddress(
+            function.function_virtual_address(), module.address_start(), module.load_bias(),
+            module.executable_segment_offset());
+    for (const LinuxMemoryMapping& map : file_path_maps) {
+      ORBIT_CHECK(map.pathname() == function.file_path());
+      if (map.start_address() <= function_absolute_address &&
+          function_absolute_address < map.end_address()) {
```
Comment (dpallotti) [inline src/LinuxTracing/LinuxTracingUtils.cpp:256]: Maybe here I should also check that the function's offset in the file is contained in the file offset range of the map?

Comment (florian-kuebler) [reply src/LinuxTracing/LinuxTracingUtils.cpp:256]: Intuitively, I would have said that the existing check is sufficient, but it would probably not hurt to also check the offset. Why do you think that check would be needed?

Comment (dpallotti) [reply src/LinuxTracing/LinuxTracingUtils.cpp:256]: I don't think it's needed, conceptually this should be enough. The reason for using the absolute addresses is to avoid false negatives as explained in the comment above, but ultimately it's still about file mappings and file offsets, so I considered this for completeness, but I'm not sure.

File: src/LinuxTracing/LinuxTracingUtils.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (danielfenner) [reply src/LinuxTracing/LinuxTracingUtils.cpp:235]: > and part of the module loaded with an anonymous map (but uprobes needs a file mapping).

~~That's the case I had in mind. Let's say there is a function at file offset 0x0500 in the example below. `FunctionIsInFileMapping` would return true since the function is "accidentally" in the non-executable file mapping (but gets executed from the executable anonymous mapping). So the function is inaccessible to uprobes but we don't generate a warning.~~

edit: And now I understood it: The `function_absolute_address` would be in the anonymous mapping and so `FunctionIsInFileMapping` would return false and that is exactly the case you want to exclude here.

Thanks for the explanations - makes sense now!

**Review by danielfenner — APPROVED** [review]
lgtm!

File: src/LinuxTracing/LinuxTracingUtils.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dpallotti) [reply src/LinuxTracing/LinuxTracingUtils.cpp:235]: Yep, that's exactly it.

File: src/LinuxTracing/LinuxTracingUtils.cpp (hunk 1 above)
Comment (dpallotti) [reply src/LinuxTracing/LinuxTracingUtils.cpp]: I changed this function (and renamed it) to return true (if and) only if a file mapping containing the absolute address of the function exists for all occurrences of the module. Please have a look and let me know if this is what you were thinking of.

File: src/LinuxTracing/LinuxTracingUtils.cpp [hunk 5]
```diff
@@ -213,4 +222,126 @@ bool SetMaxOpenFilesSoftLimit(uint64_t soft_limit) {
...
+                                            function.file_path());
+      function_ids_to_error_messages.emplace(function.function_id(), std::move(message));
+      continue;
+    }
+
+    const std::vector<LinuxMemoryMapping>& file_path_maps = file_path_maps_it->second;
+    const std::vector<ModuleInfo>& file_path_modules = file_path_modules_it->second;
+
+    if (FunctionIsAlwaysInFileMapping(file_path_maps, file_path_modules, function)) {
+      // This function is mapped into at least one file mapping.
```
Comment (florian-kuebler) [inline src/LinuxTracing/LinuxTracingUtils.cpp:317]: The comment needs to be updated.

**Review by florian-kuebler — APPROVED** [review]
Only two minor comments. LGTM. Thanks!

File: src/LinuxTracing/LinuxTracingUtils.cpp [hunk 6]
```diff
@@ -213,4 +222,126 @@ bool SetMaxOpenFilesSoftLimit(uint64_t soft_limit) {
   return true;
 }
 
+// Check for the existence of a file mapping containing the absolute address of the function.
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply src/LinuxTracing/LinuxTracingUtils.cpp]: Done. Originally I had update the rest of the paragraph, but I rephrased now and it's cleaner.

File: src/LinuxTracing/LinuxTracingUtils.cpp (hunk 5 above)
Comment (dpallotti) [reply src/LinuxTracing/LinuxTracingUtils.cpp]: Thanks, done.
